### PR TITLE
Remove (almost) unused stage_idx

### PR DESCRIPTION
**Pull request by domenukk**
This removes the `stage_idx` parameter from mutators that nothing uses

### CODE REVIEW

File: libafl/src/mutators/token_mutations.rs [hunk 1]
```diff
@@ -1144,9 +1129,7 @@ where
         // println!("orig: {:#?} new: {:#?}", orig_cmpvals, new_cmpvals);
 
         // Compute when mutating it for the 1st time.
-        if stage_idx == 0 {
-            self.text_type = check_if_text(orig_bytes, orig_bytes.len());
-        }
+        self.text_type = Some(check_if_text(orig_bytes, orig_bytes.len()));
```
Comment (domenukk) [inline libafl/src/mutators/token_mutations.rs:1132]: @tokatoka can you please double check if that's what this is supposed to do?

Comment (tokatoka) [reply libafl/src/mutators/token_mutations.rs:1132]: not it's not the same but you can do

if self.text_type == None {
 self.text_type = Some(check_if_text(...))
}

Comment (domenukk) [reply libafl/src/mutators/token_mutations.rs:1132]: Hmm it's different, the idx 0 will run for newly scheduled testcases, while the condition won't know when a new testcase got scheduled. We'll have to look if the current_corpus_idx changed here, right?

**Comment by domenukk** [comment]
@andreafioraldi as far as I understand, a mutator can get the information by counting the amount of times it got called internally, right? The only information that's missing is when a new `fuzz_one` started(?)

File: libafl/src/mutators/token_mutations.rs (hunk 1 above)
Comment (tokatoka) [reply libafl/src/mutators/token_mutations.rs:1132]: yes you are right.
what i want to do here is to check it's text type for every new testcase.

**Comment by andreafioraldi** [comment]
yes, but i guess there are multiple other options, for instance it can have
a hash field that compares with the hash of the input, if different it
means it is a different stage so reset the counter to 0

Il mer 13 mar 2024, 13:44 Dominik Maier ***@***.***> ha
scritto:

**Comment by domenukk** [comment]
It can just get the `current_corpus_idx` - but there is a tiny change that the next `fuzz_one` will schedule the same testcase again...

File: libafl/src/mutators/token_mutations.rs (hunk 1 above)
Comment (domenukk) [reply libafl/src/mutators/token_mutations.rs]: Done, PTAL

**Comment by domenukk** [comment]
So far we lived well without these indexes, but they are needed for AFLpp's deterministic mutations. I opt for removing them and potentially adding a new / cleaned up API later if we see something is missing - but it's up to @andreafioraldi to decide :)

File: libafl/src/stages/mutational.rs [hunk 2]
```diff
@@ -123,11 +118,11 @@ where
         drop(testcase);
         mark_feature_time!(state, PerfFeature::GetInputFromCorpus);
 
-        for i in 0..num {
```
Comment (domenukk) [inline libafl/src/stages/mutational.rs:126]: We can also assign this value to `state.stage_idx`, should we?

**Comment by tokatoka** [comment]
> It can just get the current_corpus_idx - but there is a tiny change that the next fuzz_one will schedule the same testcase again...

This is okay. because this stage is never supposed to be scheduled more than once against the same input

**Comment by tokatoka** [comment]
what i want to say is that you can but you should never use like that.
because this stage is a deterministic mutation so you don't run it twice.

so you usually wrap it around IfStage(scheduled_count == 2) so that, for example, this will be run on its 3rd schedule.

**Comment by domenukk** [comment]
So I guess we are all fine with removing it then? @andreafioraldi ?

**Pull request merged**